File: MatlabCodes/Chapter 05/Fig05_03.m
function MertonProcess_paths_3d
clc;close all;

% Monte Carlo settings

NoOfPaths = 200;
NoOfSteps = 500;
T         = 5;

% Model parameters

r     = 0.05;
S0    = 100;
sigma = 0.2;
sigmaJ= 0.5;
muJ   = 0;
xiP   = 1;

% Compute ChF for the Merton model

cf = ChFForMertonModel(r,T,muJ,sigmaJ,sigma,xiP,S0);

% Define the COS method integration range

L = 8;
a = - L * sqrt(T);
b = + L * sqrt(T);

% Domain for the density f(x)

x = linspace(-2,12,250);

% Exact representation 

f_XExact = COSDensity(cf,x,2^14,a,b);

% Monte Carlo paths for the Merton process

[S,time,X] = GeneratePathsMerton(NoOfPaths,NoOfSteps,S0,T,xiP,muJ,sigmaJ,r,sigma);

figure(1)
plot(time,X,'color',[0 0.45 0.74],'linewidth',1.2)
xlabel('time')
ylabel('$$X(t)$$','interpreter','latex')
grid on

figure(2)
plot(time,S,'color',[0 0.45 0.74],'linewidth',1.2)
xlabel('time')
ylabel('$$S(t)$$','interpreter','latex')
grid on

% Martingale error

error = mean(S(:,end))-S0*exp(r*T)

% Plot densities, from MC simulation and from COS method

[yCDF,xCDF]=ksdensity(X(:,end));
figure(3)
plot(xCDF,yCDF,'-r','linewidth',2.0)
hold on
plot(x,f_XExact,'--b','linewidth',2.0)
grid on
xlabel('x')
ylabel('PDF')
legend('Monte Carlo','COS method')

%%%%%%%%%%%%%%%%%% 3D figure for X(t) %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% 3D density for X(t)

figure1 = figure;
axes1 = axes('Parent',figure1);
hold(axes1,'on');
plot(time,X(1:15,:),'linewidth',1,'color',[0 0.45 0.74])

% Grid for the densities of Merton model

Tgrid=linspace(1,T,5);

% Domain for the density f(x)

x_arg=linspace(0,max(max(X(:,:))),250);
mm=0;
for i=1:length(Tgrid)    

    % Compute ChF for the Merton model

    cf = ChFForMertonModel(r,Tgrid(i),muJ,sigmaJ,sigma,xiP,S0);
    L = 8;
    a = - L * sqrt(T);
    b = + L * sqrt(T);
    f_COS = COSDensity(cf,x_arg,2^14,a,b);
    plot3(Tgrid(i)*ones(length(x_arg),1),x_arg,f_COS,'k','linewidth',2)
    mm=max(mm,max(f_COS));
end
axis([0,Tgrid(end),0,max(max(X))])
grid on;
xlabel('t')
ylabel('X(t)')
zlabel('Merton density')
view(axes1,[-68.8 40.08]);
axis([0,T,min(min(x_arg))-1,max(max(x_arg))+1,0,mm])

%%%%%%%%%%%%%%%%%% 3D figure for S(t) %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% 3D density

figure1 = figure;
axes1 = axes('Parent',figure1);
hold(axes1,'on');
plot(time,S(1:15,:),'linewidth',1,'color',[0 0.45 0.74])

% Grid for the densities of Merton model

Tgrid=linspace(1,T,5);

% Domain for the density f(x)

x_arg=linspace(0,max(max(X(:,:))),250);
mm=0;
for i=1:length(Tgrid)    

    % Compute ChF for the Merton model

    cf = ChFForMertonModel(r,Tgrid(i),muJ,sigmaJ,sigma,xiP,S0);
    L = 8;
    a = - L * sqrt(Tgrid(i));
    b = + L * sqrt(Tgrid(i));
    f_COS = 1./exp(x_arg).* COSDensity(cf,x_arg,2^14,a,b);
    plot3(Tgrid(i)*ones(length(exp(x_arg)),1),exp(x_arg),f_COS,'k','linewidth',2)
    mm=max(mm,max(f_COS));
end
axis([0,Tgrid(end),0,max(max(S0*exp(r*T)*10))])
grid on;
xlabel('t')
ylabel('S(t)')
zlabel('Merton density')
view(axes1,[-68.8 40.08]);

function cf = ChFForMertonModel(r,tau,muJ,sigmaJ,sigma,xiP,S0)

% Term for E(exp(J)-1)

i = complex(0,1);
helpExp = exp(muJ + 0.5 * sigmaJ * sigmaJ) - 1.0;
  
% Characteristic function for Merton's model    

cf = @(u) exp(i*u*log(S0)).*exp(i * u .* (r - xiP * helpExp - 0.5 * sigma * sigma) *tau...
        - 0.5 * sigma * sigma * u.^2 * tau + xiP * tau .* ...
        (exp(i * u * muJ - 0.5 * sigmaJ * sigmaJ * u.^2)-1.0));

function f_X = COSDensity(cf,x,N,a,b)
i = complex(0,1); %assigning i=sqrt(-1)
k = 0:N-1; 
u = k * pi / (b - a);

% F_k coefficients

F_k    = 2 / (b - a) * real(cf(u) .* exp(-i * u * a));
F_k(1) = F_k(1) * 0.5; % Multiply the first term

% Final calculation

f_X = F_k * cos(u' * (x - a));

    
function [S,time,X] =GeneratePathsMerton(NoOfPaths,NoOfSteps,S0,T,xiP,muJ,sigmaJ,r,sigma)

% Empty matrices for the Poisson process and stock paths

Xp     = zeros(NoOfPaths,NoOfSteps);
X      = zeros(NoOfPaths,NoOfSteps);
W      = zeros(NoOfPaths,NoOfSteps);
X(:,1) = log(S0);
S      = zeros(NoOfPaths,NoOfSteps);
S(:,1) = S0;
dt        = T/NoOfSteps;

% Random noise

Z1 = random('poisson',xiP*dt,[NoOfPaths,NoOfSteps]);
Z2 = random('normal',0,1,[NoOfPaths,NoOfSteps]);
J  = random('normal',muJ,sigmaJ,[NoOfPaths,NoOfSteps]);

% Creation of the paths

% Expectation E(exp(J))

EeJ = exp(muJ + 0.5*sigmaJ^2);
time = zeros([NoOfSteps+1,1]);
for i=1:NoOfSteps  
    if NoOfPaths>1
        Z2(:,i) = (Z2(:,i)-mean(Z2(:,i)))/std(Z2(:,i));
    end
    Xp(:,i+1) = Xp(:,i) + Z1(:,i);
    W(:,i+1)  = W(:,i)  + sqrt(dt)* Z2(:,i);
        
    X(:,i+1)  = X(:,i) + (r- xiP*(EeJ-1)-0.5*sigma^2)*dt +  sigma* (W(:,i+1)-W(:,i)) + J(:,i).*(Xp(:,i+1)-Xp(:,i));
    S(:,i+1) = exp(X(:,i));
    time(i+1) = time(i) + dt;
end

File: PythonCodes/Chapter 05/Fig05_03.py
#%%
"""
Created on Feb 08 2019
Merton process and the corresponding density obtained with the COS method
@author: Lech A. Grzelak
"""
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from mpl_toolkits import mplot3d

# Set i= imaginary number

i   = np.complex(0.0,1.0)


def GeneratePathsMerton(NoOfPaths,NoOfSteps,S0, T,xiP,muJ,sigmaJ,r,sigma):    

    # Create empty matrices for the Poisson process and compensated Poisson process

    X = np.zeros([NoOfPaths, NoOfSteps+1])
    S = np.zeros([NoOfPaths, NoOfSteps+1])
    time = np.zeros([NoOfSteps+1])
                
    dt = T / float(NoOfSteps)
    X[:,0] = np.log(S0)
    S[:,0] = S0
    
    # Expectation E(e^J) for J~N(muJ,sigmaJ^2)

    EeJ = np.exp(muJ + 0.5*sigmaJ*sigmaJ)
    ZPois = np.random.poisson(xiP*dt,[NoOfPaths,NoOfSteps])
    Z = np.random.normal(0.0,1.0,[NoOfPaths,NoOfSteps])
    J = np.random.normal(muJ,sigmaJ,[NoOfPaths,NoOfSteps])
    for i in range(0,NoOfSteps):

        # Making sure that samples from a normal have mean 0 and variance 1

        if NoOfPaths > 1:
            Z[:,i] = (Z[:,i] - np.mean(Z[:,i])) / np.std(Z[:,i])

        # Making sure that samples from a normal have mean 0 and variance 1

        X[:,i+1]  = X[:,i] + (r - xiP*(EeJ-1) - 0.5*sigma*sigma)*dt +sigma*np.sqrt(dt)* Z[:,i]\
                    + J[:,i] * ZPois[:,i]
        time[i+1] = time[i] +dt
        
    S = np.exp(X)
    paths = {"time":time,"X":X,"S":S}
    return paths

def ChFForMertonModel(r,tau,muJ,sigmaJ,sigma,xiP, S0):

    # Term for E(exp(J)-1)

    helpExp = np.exp(muJ + 0.5 * sigmaJ * sigmaJ) - 1.0
    
    # Characteristic function for Merton's model    

    cf = lambda u: np.exp(i*u*np.log(S0)) * np.exp(i * u * (r - xiP * helpExp - 0.5 * sigma * sigma) *tau \
        - 0.5 * sigma * sigma * u * u * tau + xiP * tau * \
        (np.exp(i * u * muJ - 0.5 * sigmaJ * sigmaJ * u * u)-1.0))
    return cf 

def COSDensity(cf,x,N,a,b):
    i = np.complex(0.0,1.0) #assigning i=sqrt(-1)
    k = np.linspace(0,N-1,N)
    u = np.zeros([1,N])
    u = k * np.pi / (b-a)
        
    # F_k coefficients

    F_k    = 2.0 / (b - a) * np.real(cf(u) * np.exp(-i * u * a));
    F_k[0] = F_k[0] * 0.5; # adjustment for the first term
    
    # Final calculation

    f_X = np.matmul(F_k , np.cos(np.outer(u, x - a )))
        
    # We output only the first row

    return f_X

def mainCalculation():
    NoOfPaths = 250
    NoOfSteps = 500
    T      = 5.0
    xiP    = 1.0
    muJ    = 0.0
    sigmaJ = 0.5
    sigma  = 0.2
    S0     = 100.0
    r      = 0.05
    Paths = GeneratePathsMerton(NoOfPaths,NoOfSteps,S0, T,xiP,muJ,sigmaJ,r,sigma)
    timeGrid = Paths["time"]
    X = Paths["X"]
    S = Paths["S"]
           
    plt.figure(1)
    plt.plot(timeGrid, np.transpose(X))   
    plt.grid()
    plt.xlabel("time")
    plt.ylabel("X(t)")
    
    plt.figure(2)
    plt.plot(timeGrid, np.transpose(S))   
    plt.grid()
    plt.xlabel("time")
    plt.ylabel("S(t)")
    
    # Configuration for the COS method

    L = 8.0
    a = -L * np.sqrt(T)
    b = +L * np.sqrt(T)
    
    # Define domain for density

    x = np.linspace(-2.0,12.0,250)
    
    # Comparison of the density from Monte Carlo and the COS method

    cf = ChFForMertonModel(r, T, muJ, sigmaJ,sigma,xiP, S0)
    f_xCOS = COSDensity(cf,x,1000,a,b)                   
    plt.figure(3)
    sns.distplot(X[:,-1], hist = False, kde = True, rug = False, color = 'darkblue',\
                 kde_kws={'linewidth': 3}, rug_kws={'color': 'black'}) 
    plt.plot(x,f_xCOS,'--r',linewidth=2.0)
    plt.grid()
    plt.xlabel('x')
    plt.ylabel('PDF')
    plt.title('comparison of PDF from MC and COS')

    # 3D graph for X(t) for paths and density

    plt.figure(4)
    ax = plt.axes(projection='3d')
    zline = np.zeros([len(timeGrid),1])

    # Plot paths

    n = 10
    for i in range(0,n,1):
        y1 = np.squeeze(np.transpose(X[i,:]))
        x1 = timeGrid
        z1 = np.squeeze(zline)
        ax.plot3D(x1, y1, z1, 'blue')
    ax.view_init(50, -170)
    
    Ti = np.linspace(0.5,T,5)
    
    y1 = np.linspace(0,np.max(np.max(X[:,:])),250)
    for ti in Ti:
        cf = ChFForMertonModel(r, ti, muJ, sigmaJ,sigma,xiP, S0)
        f_xCOS = COSDensity(cf,y1,1000, -L * np.sqrt(T), L * np.sqrt(T))                   
        x1 = np.zeros([len(y1),1]) + ti
        z1 = f_xCOS
        ax.plot3D(x1, y1, z1, 'red')    
    plt.xlabel('Time')
    plt.ylabel('X(t)')
        
    # 3D graph for S(t) for paths and density

    plt.figure(5)
    ax = plt.axes(projection='3d')
    zline = np.zeros([len(timeGrid),1])

    # Plot paths

    n = 10
    for i in range(0,n,1):
        y1 = np.squeeze(np.transpose(S[i,:]))
        x1 = timeGrid
        z1 = np.squeeze(zline)
        ax.plot3D(x1, y1, z1, 'blue')
    ax.view_init(50, -170)
           
    y1 = np.linspace(0,6.2,250)
    for ti in Ti:
        cf = ChFForMertonModel(r, ti, muJ, sigmaJ,sigma,xiP, S0)
        f_xCOS = COSDensity(cf,y1,1000, -L * np.sqrt(T), L * np.sqrt(T))                   
        x1 = np.zeros([len(y1),1]) + ti
        z1 = 1.0/np.exp(y1)*f_xCOS
        ax.plot3D(x1, np.exp(y1), z1, 'red')    
    plt.xlabel('Time')
    plt.ylabel('S(t)')
    
mainCalculation()
